conda_build/render.py

```diff
@@ -329,6 +330,70 @@ def add_upstream_pins(m, permit_unsatisfiable_variants, exclude_pattern):
     return build_unsat, host_unsat
 
 
+class _Constraint(object):
+    def __init__(self, lower_bound=None, upper_bound=None, exact=None):
+        self.lower_bound = lower_bound
+        self.upper_bound = upper_bound
+        self.exact = exact
+
+    def __str__(self):
+        pin = ""
+        if self.exact:
+            pin = self.exact
+        elif self.lower_bound:
+            pin = ">=" + self.lower_bound.norm_version + (
+                ',<' + self.upper_bound.norm_version) if self.upper_bound else ""
+        elif self.upper_bound:
+            pin = "<" + self.upper_bound.norm_version
+        return pin
+
+
+def _minimize_constraints(one, two):
+    if one.lower_bound and two.lower_bound:
+        lower_bound = max(VersionOrder(one.lower_bound), VersionOrder(two.lower_bound))
+    else:
+        lower_bound = one.lower_bound if one.lower_bound else two.lower_bound
+    if one.upper_bound and two.upper_bound:
+        upper_bound = min(VersionOrder(one.upper_bound), VersionOrder(two.upper_bound))
+    else:
+        upper_bound = one.upper_bound if one.upper_bound else two.upper_bound
+    if one.exact == two.exact or any(x.exact is None for x in (one, two)):
+        exact = one.exact if one.exact else two.exact
+    elif one.exact != two.exact:
+        raise ValueError("Incompatible exact pins: {} and {}".format(one, two))
+    return _Constraint(lower_bound=lower_bound, upper_bound=upper_bound, exact=exact)
+
+
+def _simplify_to_tightest_constraint(metadata):
+    requirements = metadata.meta.get('requirements', {})
+    default_constraint = _Constraint(None, None, None)
+    # collect deps on a per-section basis
+    for section in 'build', 'host', 'run':
+        deps = requirements.get(section, [])
+        deps_dict = defaultdict(_Constraint)
+        for dep in deps:
+            spec_parts = dep.split(' ')
+            name = spec_parts[0]
+            if len(spec_parts) == 1:
+                constraint = _Constraint(None, None, None)
+            elif len(spec_parts) == 3:
+                constraint = _Constraint(None, None, " ".join(spec_parts[1:]))
+            else:
+                lower_bound = re.search(r'>=?(.*?(?=,|\Z))', spec_parts[1])
+                if lower_bound:
+                    lower_bound = lower_bound.group(1)
+                upper_bound = re.search(r'<=?(.*?(?=,|\Z))', spec_parts[1])
+                if upper_bound:
+                    upper_bound = upper_bound.group(1)
+                constraint = _Constraint(lower_bound, upper_bound, None)
```

Comment:
If the regexes fail to match then no constraints are enforced.  Will this break something like `numpy ==1.11`?

Reply:
yeah, I have probably missed several cases.  ``1.21.*`` specs, too.  Please post them here or add them to the tests - I think you can just push to my branch if you want.